### PR TITLE
[debug dump util] Match Infrastructure

### PR DESCRIPTION
**Pull request by vivekreddynv**
**DEBUG DUMP UTIL PR SEQ: 1**

#### What I did

HLD for Dump Utility: [HLD](https://github.com/Azure/SONiC/pull/789).
* Added the Logic for Match Infrastructure along with corresponding unit tests. 

**Note: Before merging other PR's, please merge this first**

For More Info on MatchRequest and MatchEngine, Check this section in the HLD:
[MatchInfra](https://github.com/Azure/SONiC/blob/fda816c24570df33f253e525bb3fee7ae72fa431/doc/Dump-Utility.md#24-Match-Infrastructure)

#### How to verify it

```
tests/dump_tests/match_engine_test.py::TestMatchRequestValidation::test_bad_key_pattern PASSED [ 82%]
tests/dump_tests/match_engine_test.py::TestMatchRequestValidation::test_bad_request PASSED [ 83%]
tests/dump_tests/match_engine_test.py::TestMatchRequestValidation::test_invalid_combination PASSED [ 83%]
tests/dump_tests/match_engine_test.py::TestMatchRequestValidation::test_invalid_db PASSED [ 83%]
tests/dump_tests/match_engine_test.py::TestMatchRequestValidation::test_just_keys_return_fields_compat PASSED [ 83%]
tests/dump_tests/match_engine_test.py::TestMatchRequestValidation::test_no_file PASSED [ 83%]
tests/dump_tests/match_engine_test.py::TestMatchRequestValidation::test_no_source PASSED [ 83%]
tests/dump_tests/match_engine_test.py::TestMatchRequestValidation::test_no_table PASSED [ 83%]
tests/dump_tests/match_engine_test.py::TestMatchRequestValidation::test_no_value PASSED [ 83%]
tests/dump_tests/match_engine_test.py::TestMatchRequestValidation::test_return_fields_bad_format PASSED [ 83%]
tests/dump_tests/match_engine_test.py::TestMatchRequestValidation::test_vague_source PASSED [ 83%]
tests/dump_tests/match_engine_test.py::TestMatchRequestValidation::test_valid_match_request PASSED [ 83%]
tests/dump_tests/match_engine_test.py::TestMatchEngine::test_field_value_match PASSED [ 84%]
tests/dump_tests/match_engine_test.py::TestMatchEngine::test_field_value_match_list_type PASSED [ 84%]
tests/dump_tests/match_engine_test.py::TestMatchEngine::test_field_value_no_match PASSED [ 84%]
tests/dump_tests/match_engine_test.py::TestMatchEngine::test_file_source PASSED [ 84%]
tests/dump_tests/match_engine_test.py::TestMatchEngine::test_file_source_with_key_ptrn PASSED [ 84%]
tests/dump_tests/match_engine_test.py::TestMatchEngine::test_file_source_with_not_only_return_keys PASSED [ 84%]
tests/dump_tests/match_engine_test.py::TestMatchEngine::test_for_no_key_match PASSED [ 84%]
tests/dump_tests/match_engine_test.py::TestMatchEngine::test_for_no_match PASSED [ 84%]
tests/dump_tests/match_engine_test.py::TestMatchEngine::test_just_keys_false PASSED [ 84%]
tests/dump_tests/match_engine_test.py::TestMatchEngine::test_key_pattern_complex PASSED [ 84%]
tests/dump_tests/match_engine_test.py::TestMatchEngine::test_key_pattern_wildcard PASSED [ 84%]
tests/dump_tests/match_engine_test.py::TestMatchEngine::test_match_entire_list PASSED [ 85%]
tests/dump_tests/match_engine_test.py::TestMatchEngine::test_return_fields_with_all_filtering PASSED [ 85%]
tests/dump_tests/match_engine_test.py::TestMatchEngine::test_return_fields_with_field_value_filtering PASSED [ 85%]
tests/dump_tests/match_engine_test.py::TestMatchEngine::test_return_fields_with_key_filtering PASSED [ 85%]
tests/dump_tests/match_engine_test.py::TestMatchEngine::test_return_keys PASSED [ 85%]
```

### CODE REVIEW

**Review by SuvarnaMeenakshi — CHANGES_REQUESTED** [review]
Please added unit test for namespace DBs as well.

**Comment by vivekreddynv** [comment]
> Please added unit test for namespace DBs as well.

I've added those tests in the top level CLI.
```
tests/dump_tests/dump_state_test.py::TestDumpStateMultiAsic::test_default_namespace PASSED [ 82%]
tests/dump_tests/dump_state_test.py::TestDumpStateMultiAsic::test_namespace_asic0 PASSED [ 82%]
tests/dump_tests/dump_state_test.py::TestDumpStateMultiAsic::test_namespace_asic1 PASSED [ 82%]
tests/dump_tests/dump_state_test.py::TestDumpStateMultiAsic::test_invalid_namespace PASSED [ 82%]
```
[https://github.com/Azure/sonic-utilities/blob/4c3b775b03e090a2e0c0cb60b283c3ea143d88fc/tests/dump_tests/dump_state_test.py#L184](https://github.com/Azure/sonic-utilities/blob/4c3b775b03e090a2e0c0cb60b283c3ea143d88fc/tests/dump_tests/dump_state_test.py#L184)

File: dump/match_infra.py [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+        
+        if not isinstance(self.return_fields, list):
+            return EXCEP_DICT["BAD_FORMAT_RE_FIELDS"]
+        
+        if not self.just_keys and self.return_fields:
+            return EXCEP_DICT["JUST_KEYS_COMPAT"]
+        
+        if self.field and not self.value:
+            return EXCEP_DICT["NO_VALUE"]
+
```
[the remark this answers is not in the document]
Comment (vivekreddynv) [reply dump/match_infra.py:90]: Check added

File: tests/dump_tests/match_engine_test.py [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+    def test_vague_source(self):  
+        self.assertRaisesWithMessage(EXCEP_DICT["SRC_VAGUE"], MatchRequest, db="CONFIG_DB", file="/etc/sonic/copp_cfg.json")
+    
+    def test_no_file(self):  
+        self.assertRaisesWithMessage(EXCEP_DICT["FILE_R_EXEP"], MatchRequest, file=os.path.join(test_path, "random_db.json"))
+
+    def test_invalid_db(self):  
+        self.assertRaisesWithMessage(EXCEP_DICT["INV_DB"], MatchRequest, db="CONFIGURATION_DB")
+        
+    def test_bad_key_pattern(self):
```
[the remark this answers is not in the document]
Comment (vivekreddynv) [reply tests/dump_tests/match_engine_test.py:57]: Yeah, there is. For MatchEngine tests i'm using the data in the mock_tables/ folder

File: dump/match_infra.py [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+    "NO_SRC": "Either one of db or file in the request should be non-empty",
+    "NO_TABLE": "No 'table' name provided",
+    "NO_KEY": "'key_pattern' cannot be empty",
+    "NO_VALUE" : "Field is provided, but no value is provided to compare with", 
+    "SRC_VAGUE": "Only one of db or file should be provided",
+    "CONN_ERR" : "Connection Error",
+    "JUST_KEYS_COMPAT": "When Just_keys is set to False, return_fields should be empty",
+    "BAD_FORMAT_RE_FIELDS": "Return Fields should be of list type",
+    "NO_ENTRIES": "No Keys found after applying the filtering criteria",
+    "FILE_R_EXEP": "Exception Caught While Reading the json cfg file provided"
```
[the remark this answers is not in the document]
Comment (vivekreddynv) [reply dump/match_infra.py]: Added

File: tests/dump_tests/match_engine_test.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (SuvarnaMeenakshi) [reply tests/dump_tests/match_engine_test.py:57]: Got it, the existing mock_tables is getting used.

**Comment by liat-grozovik** [comment]
@SuvarnaMeenakshi could you please review following unit test added?

**Review by SuvarnaMeenakshi — APPROVED** [review]
LGTM

**Review by liat-grozovik — APPROVED** [review]

**Comment by vivekreddynv** [comment]
Request for 202106

**Pull request merged**